WWSTCERT-10701 Add Sonoff SNZB-01M Smart Scene Button into zigbee-button.

Oniums opened the pull request:
Check all that apply

# Type of Change

- [x] WWST Certification Request
     - If this is your first time contributing code:
          - [x] I have reviewed the README.md file
          - [x] I have reviewed the CODE_OF_CONDUCT.md file
          - [x] I have signed the CLA
     - [x] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [ ] Bug fix
- [x] New feature
- [ ] Refactor

# Checklist

- [x] I have performed a self-review of my code
- [x] I have commented my code in hard-to-understand areas
- [x] I have verified my changes by testing with a device or have communicated a plan for testing
- [ ] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change

This PR adds support for the **Sonoff SNZB-01M Smart Scene Button** to the existing zigbee-button driver.

### Changes made:
1. **Added device fingerprint** in `fingerprints.yml` to recognize the SONOFF SNZB-01M device
2. **Created device profile** `sonoff-buttons-battery.yml` with 4 button components and battery capability
3. **Implemented custom handler** in `src/zigbee-multi-button/sonoff/init.lua` to handle SONOFF-specific Zigbee communication
4. **Added device support** to the multi-button matching matrix with support for 4 button actions: pushed, double, held, pushed_3x
5. **Created test file** `test_sonoff_button.lua` for verification purposes

# Summary of Completed Tests

### Device Testing:
- ✅ **Device pairing**: Successfully paired SONOFF SNZB-01M with SmartThings Hub
- ✅ **Button actions**: Verified all 4 button types work correctly:
  - Single press (pushed)
  - Double press (double) 
  - Long press (held)
  - Triple press (pushed_3x)
- ✅ **Battery reporting**: Confirmed battery percentage is correctly displayed in SmartThings app
- ✅ **Endpoint mapping**: Verified button events are properly mapped to correct button components (button1-button4)

### Code Testing:
- ✅ **Fingerprint recognition**: Device correctly identified during pairing process
- ✅ **Profile loading**: Device profile loaded successfully with all 4 button components
- ✅ **Event handling**: Custom SONOFF handler processes Zigbee messages correctly
- ✅ **Integration**: No conflicts with existing button drivers

### Compatibility:
- ✅ **SmartThings Hub compatibility**: Tested on SmartThings Hub v3
- ✅ **Driver integration**: Successfully integrates with existing zigbee-button driver framework
- ✅ **No regression**: Existing button devices continue to function normally

Oniums committed:
Add Sonoff SNZB-01M Smart Scene Button into zigbee-button.

pegor-karoglanian commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/sonoff/init.lua:
```diff
@@ -0,0 +1,87 @@
+-- Copyright 2024 SmartThings
```
Update date

Oniums committed:
modify Copyright Date

Oniums committed:
modify test_sonoff.py fix PR error

greens commented on drivers/SmartThings/zigbee-button/src/test/test_sonoff_button.lua:
```diff
@@ -0,0 +1,261 @@
+-- Copyright 2022 SmartThings
```
date

Oniums committed:
Update copyright year from 2022 to 2026,remove remove whitespace-only lines 

Update copyright year from 2022 to 2026,remove remove whitespace-only lines

Oniums committed:
remove log.info

remove log.info

greens commented on drivers/SmartThings/zigbee-button/profiles/sonoff-buttons-battery.yml:
to avoid duplication here, I'd recommend you use the existing profile `four-buttons-without-main-button`

greens commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/supported_values.lua:
```diff
@@ -130,6 +130,13 @@ local devices = {
+      { mfr = "SONOFF", model = "SNZB-01M" }
+    },
+    SUPPORTED_BUTTON_VALUES = { "pushed", "double", "held", "pushed_3x" },
+    NUMBER_OF_BUTTONS = 4
```
Since you overwrote `added` in your handler (because you did not include the `button` capability on your `main` component, this info is never used.

cbaumler replied:
@Oniums, I believe `SONOFF_BUTTON_4` is still unused and can be removed. The parent zigbee-multi-button added handler is the only place that consumes this table, and the matched SONOFF child should take over dispatch before that handler runs. If the SONOFF sub-driver needs custom button metadata initialization, put an added handler in the SONOFF sub-driver; otherwise remove this dead config entry.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=2510) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=2510) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Oniums<br/>:x: West_Zhao<br/><hr/>**West_Zhao** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=2510) it.</sub>

Oniums committed:
Use defaults for Sonoff battery and added lifecycle

- Remove custom battery attribute handler and added lifecycle in Sonoff subdriver
- Align Sonoff tests with default added behavior and keep battery report coverage

greens commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/sonoff/init.lua:
```diff
@@ -0,0 +1,61 @@
+      log.warn("Unknown button component: " .. button_name)
+    end
+  else
+    log.warn("Unknown event value: " .. tostring(attr_val))
```
remove these log statements

Oniums committed:
remove log

Oniums commented:
Is the rebase based on this PR? If I submit a new PR, will it cause any issues if it doesn’t match the PR used for the Samsung certification application?

greens commented:
@Oniums you should be able to set your fork's upstream to our main repo, update the main branch there, then rebase your changes on your local branch onto `main`:

In your fork's directory:
```
git remote add upstream git@github.com:SmartThingsCommunity/SmartThingsEdgeDrivers.git
git checkout upstream/main
git pull
git checkout main
git rebase upstream/main
```

Then you should be able to resolve the conflicts and `git push --force` up your changes.

Oniums committed:
Fix Sonoff multi-button sub-driver registration and test execution

KKlimczukS reviewed (changes requested):
Please update copyright sections

Oniums committed:
update copyright sections for sonoff multi-button

cjswedes reviewed (commented):
LGTM, just note the suggestion to reconsider adding the button capability to the main component of the device and emitting events for it when any button is affected.

cbaumler commented:
Hello @Oniums, in addition to @cjswedes's feedback about making sure the device is working the way you expect, there are also some test failures and merge conflicts that need to be addressed before this PR can be merged.

Oniums committed:
Merge branch 'SmartThingsCommunity:main' into main

Oniums committed:
fix: address review feedback for Sonoff SNZB-01M button support

- Reuse existing four-buttons-battery profile instead of custom sonoff-buttons-battery
- Mirror button events to main component via emit_event for device list visibility
- Update test profile reference and add main component event assertions
- Remove unused sonoff-buttons-battery.yml profile

Oniums committed:
fix: resolve Sonoff SNZB-01M unit test failures

- replace unsupported build_custom_report_attribute usage with an explicit custom Zigbee attribute report builder
- add missing main component expectations for the added lifecycle test
- add min_api_version guards to Sonoff test cases
- keep Sonoff button event assertions aligned with the four-buttons-battery profile behavior

cbaumler commented:
Hello @Oniums, it looks like this PR branch is out of date with main. Please update it with the latest main and push the result so it can be merged.

Oniums committed:
Merge branch 'SmartThingsCommunity:main' into main

Oniums committed:
Merge branch 'main' into main

KKlimczukS commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/sonoff/can_handle.lua:
```diff
@@ -0,0 +1,16 @@
+  return false
+end
+
+return sonoff_can_handle
```
@Oniums

Oniums committed:
Merge branch 'main' into main

cbaumler reviewed (commented):
@Oniums I added one more comment regarding `supported_values.lua`. In addition, there are still some nit comments that need to be addressed. Otherwise, the PR looks good to me.

Oniums committed:
Merge branch 'main' into main

Oniums committed:
Removed the SONOFF entry from the generic supported values table and moved its button metadata initialization into the SONOFF sub-driver.

KKlimczukS commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/sonoff/can_handle.lua:
```diff
@@ -0,0 +1,16 @@
+  return false
+end
+
+return sonoff_can_handle
```
```suggestion
return sonoff_can_handle

```

KKlimczukS commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/sonoff/fingerprints.lua:
```diff
@@ -0,0 +1,8 @@
+  { mfr = "SONOFF", model = "SNZB-01M" }
+}
+
+return SONOFF_FINGERPRINTS
```
```suggestion
return SONOFF_FINGERPRINTS

```

KKlimczukS commented on drivers/SmartThings/zigbee-button/src/zigbee-multi-button/sonoff/init.lua:
```diff
@@ -0,0 +1,62 @@
+  can_handle = require("zigbee-multi-button.sonoff.can_handle")
+}
+
+return sonoff_handler
```
```suggestion
return sonoff_handler

```

KKlimczukS commented on drivers/SmartThings/zigbee-button/src/test/test_sonoff_button.lua:
```diff
@@ -0,0 +1,278 @@
+  }
+)
+
+test.run_registered_tests()
```
```suggestion
test.run_registered_tests()

```

Oniums committed:
Apply suggestion from @KKlimczukS

Co-authored-by: Konrad K <33450498+KKlimczukS@users.noreply.github.com>

Oniums committed:
Apply suggestion from @KKlimczukS

Co-authored-by: Konrad K <33450498+KKlimczukS@users.noreply.github.com>

Oniums committed:
Apply suggestion from @KKlimczukS

Co-authored-by: Konrad K <33450498+KKlimczukS@users.noreply.github.com>

Oniums committed:
Apply suggestion from @KKlimczukS

Co-authored-by: Konrad K <33450498+KKlimczukS@users.noreply.github.com>

cbaumler reviewed (approved):
LGTM. Thanks for making the changes! 

Please update the PR branch with the latest main and push the result, and we should be able to merge.

Oniums committed:
Merge branch 'main' into main

KKlimczukS reviewed (approved):
lgtm
